### PR TITLE
Add niryo ned robot

### PR DESCRIPTION
**Pull request by NicolasGuy97**
**Description**
First pull request (draft) to add Ned robot from Niryo into Webots. 

**Related Issues**
For the moment the video is not done yet, and the icons are not the final one but our marketing team is working on it. The goal of this pull request is to see if the ned.proto model, the controllers and the world are OK. 

**Tasks**
Add the list of tasks of this PR.
  - [ ] Check size of the proto, textures and meshes files.
  - [ ] Check if the world is OK.
  - [ ] Check of the controllers (C and Python) are OK.
  - [ ] Check if the documentation is OK.

**Documentation**
https://cyberbotics.com/doc/guide/robots?version=NicolasGuy97:niryo_ned_robot
https://cyberbotics.com/doc/guide/ned?version=NicolasGuy97:niryo_ned_robot

### CODE REVIEW

File: projects/robots/niryo/ned/controllers/ned_c_controller/ned_c_controller.c
Comment (omichel) [inline projects/robots/niryo/ned/controllers/ned_c_controller/ned_c_controller.c:215]:
```suggestion
```

File: projects/robots/niryo/ned/controllers/ned_c_controller/ned_c_controller.c
Comment (omichel) [inline projects/robots/niryo/ned/controllers/ned_c_controller/ned_c_controller.c:40]:
```suggestion
  printf("Move joint_1 --> A and Z\n");
```
etc.

File: projects/robots/niryo/ned/controllers/ned_python_controller/ned_python_controller.py
Comment (omichel) [inline projects/robots/niryo/ned/controllers/ned_python_controller/ned_python_controller.py:19]:
```suggestion
print('Name of the robot: ' + robot_name)
```

**Comment by NicolasGuy97** [comment]
Hello Olivier, I did a second update on the PR, looks better but not done yet. I still have issues with the pictures but it's normal, I am waiting for our marketing team to send me the appropriates ones. Concerning the world, I got rid of the other Ned robots. Let me know what do you think about it now. 

The two controllers (Python and C) are still not respecting the PEP and Clang format but it's on a good way.

Thank you for reading

**Comment by omichel** [comment]
Thank you for the update! You can press the "Ready for review" button when you believe it is ready for review. Or if you need some intermediate review, simply let me know by posting a message here and I will review the PR again.

**Comment by NicolasGuy97** [comment]
I think the robot part is ok, I am now working on the documentation part, how can I generate:
  -images/robots/ned/ned.wbt.thumbnail.jpg
  -images/robots/ned/model.thumbnail.png

Thanks

**Comment by omichel** [comment]
> how can I generate:
-images/robots/ned/ned.wbt.thumbnail.jpg
-images/robots/ned/model.thumbnail.png

You can simply run the [generate_thumbnails.py](https://github.com/cyberbotics/webots/blob/master/docs/generate_thumbnails.py) python script and it should generate these two files automatically.

**Comment by NicolasGuy97** [comment]
Hello Olivier, I think I am good for this PR, I am just waiting for the video to be done (probably in the next days) and I will change this PR to "Ready For Review"

Have a good day

**Review by omichel — COMMENTED** [review]
The `.idea` folder should be removed from this PR as it is specific to some IDE which we do not support.

**Review by omichel — COMMENTED** [review]
The binary controllers, e.g, ned_c_controller should not be committed. Instead, it should be ignored (added to the local .gitignore file)

File: projects/robots/niryo/ned/controllers/ned_python/ned_python.py
Comment (omichel) [inline projects/robots/niryo/ned/controllers/ned_python/ned_python.py:218]:
```suggestion
```

File: projects/robots/niryo/ned/controllers/ned_python/ned_python.py
Comment (omichel) [inline projects/robots/niryo/ned/controllers/ned_python/ned_python.py:282]:
```suggestion
            print("Demonstrator: Move Joints")
```

File: projects/robots/niryo/ned/controllers/ned_python/ned_python.py
Comment (omichel) [inline projects/robots/niryo/ned/controllers/ned_python/ned_python.py:286]:
```suggestion
            print("Demonstrator: Pick And Place")
```

**Review by omichel — CHANGES_REQUESTED** [review]
The `Ned.png`, `icon.png`, `model.png` and `model.thumbnail.png` images contains red dots in the background, could you please remove them?
Moreover, the world file still use a copy of the content of the proto in a Robot node, rather it should use the Ned proto and not the Robot node.

**Comment by omichel** [comment]
Please use the "Commit suggestion" on the GitHub web interface to perform the proposed changes if you agree with them (and pull the changes to your local version with "git pull") rather than applying the changes locally and pushing them. This simplifies the review process for me and should also be easier for you.

**Comment by omichel** [comment]
> Can we publish it on our Youtube channel and put the link on the Ned.md ?

Yes, that's the way to go.

File: projects/robots/niryo/ned/controllers/ned_python/ned_python.py
Comment (omichel) [inline projects/robots/niryo/ned/controllers/ned_python/ned_python.py:77]: You should apply this to all the occurrences below this one as well:
```suggestion
    if robot.step(1500) == -1:
        return
```

**Review by omichel — CHANGES_REQUESTED** [review]
The ned binary controller file `projects/robots/niryo/ned/controllers/ned/ned` should not be committed on the repository. Please delete it.
Also the `ned.wbt` world file contains a huge Robot node for the Ned robot. Instead it should use the proto.

File: docs/guide/ned.md
Comment (omichel) [inline docs/guide/ned.md:32]:
```suggestion
#### ned.wbt
```

**Review by omichel — CHANGES_REQUESTED** [review]

File: docs/guide/ned.md
Comment (NicolasGuy97) [reply docs/guide/ned.md]: I just changed the readme and added the youtube video link

File: projects/robots/niryo/ned/controllers/ned_python/ned_python.py
Comment (NicolasGuy97) [reply projects/robots/niryo/ned/controllers/ned_python/ned_python.py]: Did it for all the occurences

Comment (NicolasGuy97) [reply projects/robots/niryo/ned/controllers/ned_python/ned_python.py]: I also put the robot proto and not the robot node this time

**Comment by NicolasGuy97** [comment]
Good afternoon Olivier, I did the final modifications for this PR, I put the link of our video on the the ReadMe and correct the errors you targeted. Let me know if I need to do other modifications. 

Have a good day

**Review by omichel — DISMISSED** [review]
We are almost there.
Please remove the ned binary executable as I already requested you twice.
Also, please remove all the hidden fields from the world file: they are not needed and probably the result of saving the world after it ran.

**Comment by NicolasGuy97** [comment]
I am sorry for the ned binary executable. I tought I put it on the .gitignore but it wasn't properly set. 

I am actually learning on how to use git. It's my first big Pull Request on it but thanks to Webots I am learning a lot about it for the future.

**Comment by omichel** [comment]
No problem. It's not sufficient to list it in `.gitignore`, you also have to perform a `git rm ned` and `git commit` in the ned folder.

**Comment by omichel** [comment]
Also, I noticed that the pick and place demo doesn't work: the robot fails at picking the object... Does it work for you?

**Comment by omichel** [comment]
I believe the bounding object for the fingers of the gripper should be replaced by box primitives (or a combination of other simple primitive). In general we should not use IndexedFaceSet nodes for bounding objects as they cause instabilities.
Also, I noticed that the bounding objects for the arm are missing.

**Comment by NicolasGuy97** [comment]
Hello Olivier, I pulled your modifications and the rendering is very good. Concerning the pick and place it is not working every time due to the issues you targeted. I will try to change the collision of the URDF to put Box primitives for the jaws of the gripper. 

Concerning the video I just asked the marketing team if they can do it again with this rendering. 

Thank you for you advises

**Comment by omichel** [comment]
> I will try to change the collision of the URDF to put Box primitives for the jaws of the gripper.

It should be easier to simply edit the PROTO file with a text editor and change these bounding objects to Box primitives.

**Comment by NicolasGuy97** [comment]
Good afternoon Olivier, I modified the bounding boxes of Ned, it should be better now. I test the pick and place and it's always working on my side 

I also took the new videos with the actual rendering, we are making the youtube video as fast as possible.

**Comment by omichel** [comment]
Can you wait a little bit before making the video? One of my colleagues (@mael25) is trying to improve the appearance of the model, especially the metallic parts. He will post an updated proto file on Monday. It would be nice also to retake a screenshot with the new model for the icons and the world file image in the user guide. Hopefully, we should be able to merge this PR next week. Thank you for your patience and efforts. I believe it will be a very nice model.

**Comment by NicolasGuy97** [comment]
No problem I will wait for the new proto model and take new pictures. It is a pleasure to work with you and as I am finishing my studies doing this internship I am learning a lot with Webots. I really hope that the result will be nice at the end. 

Have a good week end

**Comment by NicolasGuy97** [comment]
Hello Olivier, I hope you had a good week end and a good week so far. I would like to know if the new .proto files would be available this week because my supervisor would like to put the Pull Request "Ready for review" this week if that's possible for you.

Thank you, have a good day.

**Comment by mael25** [comment]
> Hello Olivier, I hope you had a good week end and a good week so far. I would like to know if the new .proto files would be available this week because my supervisor would like to put the Pull Request "Ready for review" this week if that's possible for you.
> 
> Thank you, have a good day.

Hello Nicolas, I have opened a [pull request](https://github.com/NicolasGuy97/webots/pull/1) Monday directly on your repo so that you can choose if you want to accept the changes made. If you prefer, I can send you the proto file by email.

Don't hesitate if you have questions or if you want to make other improvements.

Have a good day !

**Comment by NicolasGuy97** [comment]
> > Hello Olivier, I hope you had a good week end and a good week so far. I would like to know if the new .proto files would be available this week because my supervisor would like to put the Pull Request "Ready for review" this week if that's possible for you.
> > Thank you, have a good day.
> 
> Hello Nicolas, I have opened a [pull request](https://github.com/NicolasGuy97/webots/pull/1) Monday directly on your repo so that you can choose if you want to accept the changes made. If you prefer, I can send you the proto file by email.
> 
> Don't hesitate if you have questions or if you want to make other improvements.
> 
> Have a good day !

Hello @mael25, sorry for the message, I didn't see the Pull Request you did. Thank you very much for your work on the .proto file.

Have a good day too.

**Comment by mael25** [comment]
> > > Hello Olivier, I hope you had a good week end and a good week so far. I would like to know if the new .proto files would be available this week because my supervisor would like to put the Pull Request "Ready for review" this week if that's possible for you.
> > > Thank you, have a good day.
> > 
> > 
> > Hello Nicolas, I have opened a [pull request](https://github.com/NicolasGuy97/webots/pull/1) Monday directly on your repo so that you can choose if you want to accept the changes made. If you prefer, I can send you the proto file by email.
> > Don't hesitate if you have questions or if you want to make other improvements.
> > Have a good day !
> 
> Hello @mael25, sorry for the message, I didn't see the Pull Request you did. Thank you very much for your work on the .proto file.
> 
> Have a good day too.

No problem, I could have posted a message here as well in the first place. You're welcome, it was nice working on your robot !

**Comment by mael25** [comment]
@NicolasGuy97 I have re-run the tests and they are passing again. If you don't have more changes to make to the files, you can click on the "Ready for review" button and we'll make a final review before integrating the Ned robot into Webots.

**Comment by NicolasGuy97** [comment]
> @NicolasGuy97 I have re-run the tests and they are passing again. If you don't have more changes to make to the files, you can click on the "Ready for review" button and we'll make a final review before integrating the Ned robot into Webots.

I just took new images for the doc with the new model and then I will push again. Thanks
I am now recording new videos with the new model as well.

**Comment by NicolasGuy97** [comment]
I just pushed the pictures with the new model, all the tests are checked, we are just missing the new video. I am working on it to have it as soon as possible.

**Comment by mael25** [comment]
Hello @NicolasGuy97, I have added the Web component (the simulation preview that appears on top of each robot's documentation page) for the Ned robot, you can accept the changes [here](https://github.com/NicolasGuy97/webots/pull/2).

**Comment by NicolasGuy97** [comment]
> Hello @NicolasGuy97, I have added the Web component (the simulation preview that appears on top of each robot's documentation page) for the Ned robot, you can accept the changes [here](https://github.com/NicolasGuy97/webots/pull/2).

Thank you very much @mael25 for adding the Web component. Concerning the video, it should be done on Mondays. Normally if everything is ok I should put the PR "ready for review" beginning of next week.

**Comment by NicolasGuy97** [comment]
Hello, just to let you know that the person in charge of the Webots video in our company is very sick right now and won't come back before next week. In this way, I won't be able to put this pull request "Ready for review" this week. I am sorry for the delay.

Have a good day.

**Comment by mael25** [comment]
> Hello, just to let you know that the person in charge of the Webots video in our company is very sick right now and won't come back before next week. In this way, I won't be able to put this pull request "Ready for review" this week. I am sorry for the delay.
> 
> Have a good day.

Hello, thank you for the update. No problem, we wish them a good recovery. 

Have a good day too.

**Comment by NicolasGuy97** [comment]
Hello @omichel and @mael25, the person in charge of the video finally came back, he feels better now and had time to finish the video. I just put the new link on the readme, I hope you will like it. 
 
I am now ready to put the PR "Ready for review" but before that, do you have other recommendations ? How can I launch the test again to check if everything is ok ? 

Thank you, have a good afternoon.

**Comment by mael25** [comment]
> Hello @omichel and @mael25, the person in charge of the video finally came back, he feels better now and had time to finish the video. I just put the new link on the readme, I hope you will like it.
> 
> I am now ready to put the PR "Ready for review" but before that, do you have other recommendations ? How can I launch the test again to check if everything is ok ?
> 
> Thank you, have a good afternoon.

Hello @NicolasGuy97, glad to hear the person is feeling better. The video is really nice, congratulations !

We noticed that the minPosition and maxPosition of the RotationalMotor were not set, could you add them if you have the values ? This way, if someone use another controller, it guarantees that the robot will not collide with itself. I will let you know if I find something else to change.

Thanks and have a good afternoon too.

**Comment by NicolasGuy97** [comment]
> > Hello @omichel and @mael25, the person in charge of the video finally came back, he feels better now and had time to finish the video. I just put the new link on the readme, I hope you will like it.
> > I am now ready to put the PR "Ready for review" but before that, do you have other recommendations ? How can I launch the test again to check if everything is ok ?
> > Thank you, have a good afternoon.
> 
> Hello @NicolasGuy97, glad to hear the person is feeling better. The video is really nice, congratulations!
> 
> We noticed that the minPosition and maxPosition of the `RotationalMotor` were not set, could you add them if you have the values ? This way, if someone use another controller, it guarantees that the robot will not collide with itself. I will let you know if I find something else to change.
> 
> Thanks and have a good afternoon too.

Hi @mael25, thank you very much for your answer. I just added the minPosition and the maxPosition of the rotationalMotors. Let me know if it's ok and if you see something else to update. 

Thanks

**Comment by mael25** [comment]
@NicolasGuy97 I have updated the position and zoom of the X3D animation. It was not possible to replace it directly here, so I have made a Pull Request on your repository: [https://github.com/NicolasGuy97/webots/pull/3](https://github.com/NicolasGuy97/webots/pull/3)

**Comment by NicolasGuy97** [comment]
> @NicolasGuy97 I have updated the position and zoom of the X3D animation. It was not possible to replace it directly here, so I have made a Pull Request on your repository: [NicolasGuy97#3](https://github.com/NicolasGuy97/webots/pull/3)

Hello @mael25, thank you for the update, I just accepted the Pull Request

**Review by mael25 — APPROVED** [review]
@NicolasGuy97 We have reviewed the code and everything is ok for us now, you can close this Pull Request. Thanks for your work on this robot, it looks very nice and it has been great working with you !

Have a nice day.

**Comment by omichel** [comment]
@NicolasGuy97: I just sent you an invite to become a member of Cyberbotics Committers on GitHub, so that you will be able to merge your PR as soon as the tests passed.

**Review by omichel — APPROVED** [review]
Congratulation on this very nice contribution!

**Comment by NicolasGuy97** [comment]
Thank you very much @omichel  and @mael25 for the work you have done and your help to integrate Ned on Webots. Sorry for the time spent on it. It was such a pleasure for me to work with you, I learnt a lot about how Github works and robotics thanks to you. 
I really appreciated to work on Webots, it's an amazing robot simulator. Looking forward to adding new features on it for Ned. I hope people will like it and use it. 

We are almost ready for the marketing around it on our side, can you please let me know when everything is ready for you to be able to promote Ned at the same time ? 

Have a good day, maybe we can keep in touch on LinkedIn ?

**Comment by omichel** [comment]
Since your PR was merged into the master branch, it will normally automatically included in the next nightly build of Webots that you could download starting from tomorrow at https://github.com/cyberbotics/webots/releases. We plan to make a new official release of Webots at the end of June that will include the Ned. So, we can either communicate starting from tomorrow (once we verified the Ned is properly included in the nightly build and works as expected) or wait until the official release that will happen at the end of June. If we communicate earlier than the official release, we can use the following links to point to the Ned documentation: https://cyberbotics.com/doc/guide/ned?version=master and the [youtube video](https://www.youtube.com/watch?v=diBAJY1WJPQ), which looks very nice. Cyberbotics can also use its usual communication channels (Twitter, LinkedIn, Discord) to relay the news.

**Comment by NicolasGuy97** [comment]
> Since your PR was merged into the master branch, it will be normally automatically included in the next nightly build of Webots that you could download starting from tomorrow at https://github.com/cyberbotics/webots/releases. We plan to make a new official release of Webots at the end of June that will include the Ned. So, we can either communicate starting from tomorrow (once we verified the Ned is properly included in the nightly build and works as expected) or wait until the official release that will happen at the end of June. If we communicate earlier than the official release, we can use the following links to point to the Ned documentation: https://cyberbotics.com/doc/guide/ned?version=master and the [youtube video](https://www.youtube.com/watch?v=diBAJY1WJPQ), which looks very nice. Cyberbotics can also use its usual communication channels (Twitter, LinkedIn, Discord) to relay the news.

Thanks for your answer. Looking forward to seeing Ned on your Website tomorrow. I have a meeting with the head of Marketing tomorrow to fix a date of release of the marketing around Ned on Webots. I will let you know what we are going to do as soon as I know. 

Thank you again for your work.

**Pull request merged**
